src/source/geojson_source.js

```diff
@@ -108,8 +108,11 @@ class GeoJSONSource extends Evented {
                 this.fire('error', {error: err});
                 return;
             }
-            this.fire('data', {dataType: 'source'});
-            this.fire('source.load');
+            // `update` is included here to prevent a race condition where `Style#_updateSources` is called
+            // before the TileJSON arrives. this makes sure the tiles needed are loaded once TileJSON arrives
+            // ref: https://github.com/mapbox/mapbox-gl-js/pull/4347#discussion_r104418088
+            this.fire('data', {dataType: 'source', sourceDataType: 'metadata'});
```

Comment:
In this case (GeoJSONSource), the term `'metadata'` is _slightly_ odd, since there isn't really any metadata for geojson sources.  As such, it might be more helpful to replace the comment about `update` in favor of a comment reminding us that we're firing `metadata` to notify the source cache that it's 👍 to ask us for tiles.   

(I.e., reading this source, I think it's `metadata` that might be confusing; reading the vector/raster sources, it's `update` that might be confusing)